src/rust/ide/lib/view/src/graph_editor/component/node/port/output.rs

```diff
@@ -582,43 +611,88 @@ impl OutputPortsData {
-        let network = default();
-        let frp     = Frp::new(&network);
-        let data    = OutputPortsData::new(scene,number_of_ports);
-        let data    = Rc::new(data);
-        OutputPorts {data,network,frp}.init()
+    pub fn new(scene:&Scene) -> Self {
+        let expression       = Expression::default();
+        let network          = default();
+        let id_map           = default();
+        let frp              = Frp::new(&network,&id_map);
+        let number_of_ports  = expression.output_span_tree.root_ref().leaf_iter().count();
+        let data             = OutputPortsData::new(scene,number_of_ports as u32);
+        let data             = Rc::new(data);
+        let type_map         = default();
+        let expression       = Rc::new(RefCell::new(expression));
+        let scene            = scene.clone_ref();
+        let network_internal = default();
+
+        OutputPorts {scene,data,network,frp,expression,type_map,network_internal,id_map}
     }
 
-    fn init(self) -> Self {
-        let network = &self.network;
-        let frp     = &self.frp;
+    // TODO: Implement proper sorting and remove.
+    /// Hack function used to register the elements for the sorting purposes. To be removed.
+    pub(crate) fn order_hack(scene:&Scene) {
+        let logger = Logger::new("output shape order hack");
+        component::ShapeView::<multi_port_area::Shape>::new(&logger,scene);
+        component::ShapeView::<single_port_area::Shape>::new(&logger,scene);
+    }
+
+    /// Set the expression for which output ports should be presented. Triggers a rebinding of the
+    /// internal FRP and updates the shape appearance.
+    pub fn set_expression(&self, expression:impl Into<Expression>) {
+        let expression = expression.into();
+
+        // === Update data / shapes ===
+
+        // We want to be able to match each `PortId` to a `Crumb` from the expression, including
+        // the root.
+        let number_of_ports = 1 + expression.output_span_tree.root_ref().leaf_iter().count() as u32;
+        let mut id_map      = HashMap::<PortId,span_tree::Crumbs>::from_iter(expression.output_span_tree.root_ref()
+            .leaf_iter()
+            .enumerate()
+            .map(|(index, node)| {
+                (PortId::new(index + 1), node.crumbs)
+        }));
+        id_map.insert(PortId::new(0), expression.output_span_tree.root_ref().crumbs);
+        let id_map = id_map;
+        *self.id_map.borrow_mut() = id_map;
+
+        self.data.set_port_number(number_of_ports);
+        *self.expression.borrow_mut() = expression;
+
+
+        // === Update FRP bindings ===
         let data    = &self.data;
+        let frp     = &self.frp;
 
         // Used to set and detect the end of the tweens. The actual value is irrelevant, only the
         // duration of the tween matters and that this value is reached after that time.
         const TWEEN_END_VALUE:f32 = 1.0;
 
+        // TODO memory leak from tween?
```

Comment:
Is it okay to create a new one here, or does this leak memory? If so we need to instantiate them somewhere else.

Reply:
For example in the `network_internal`.

Reply:
Why you think it will leak memory? You may be right, but I need explanation to justify it.

Reply:
Whenever we change the number of ports, we need to re-create the frp bindings of the `port_network` (formerly `network_internal`). Right now we drop the `port_network` and set up new bindings to the `network`. But the `Tween` is created in the `network` which persists. 

So the question is: will it create a new tween every time or is it fine to create a new one?

Writing it out like this I heavily suspect that it's bad and needs to be another way.

Reply:
You mean that you create animations `delay_show` and `delay_hide` ON EVERY `set_expression` right? This is of course bad - you should create them on init. There is nothing bad in the creation of animations on init and THEN connecting to them multiple times from new networks. When this network will be removed, these animations will correctly handle the dropped network (will clean unused connections). But of course, creating new nodes in the same network constantly is a memory leak